service/internal/hamaster/ha_master.go

```diff
@@ -0,0 +1,170 @@
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// We need the G8sControlPlane CR because it holds the replica count. This
+	// tells us how many masters the current setup defines and ultimately dictates
+	// the Master IDs. The system's implementation requires there only to be 1 or
+	// 3 masters.
+	var g8s infrastructurev1alpha2.G8sControlPlane
+	{
+		var list infrastructurev1alpha2.G8sControlPlaneList
+
+		err := h.k8sClient.CtrlClient().List(
+			ctx,
+			&list,
+			client.InNamespace(cr.GetNamespace()),
+			client.MatchingLabels{label.Cluster: key.ClusterID(cr)},
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(list.Items) == 0 {
+			return microerror.Mask(notFoundError)
+		}
+		if len(list.Items) > 1 {
+			return microerror.Mask(tooManyCRsError)
+		}
+
+		g8s = list.Items[0]
+	}
+
+	// We need the AWSControlPlane CR because it holds the availability zones. The
+	// state machine allows to cycle through them in a deterministic way. The
+	// system's implementation requires there only to be 1, 2 or 3 availability
+	// zones.
+	var aws infrastructurev1alpha2.AWSControlPlane
+	{
+		var list infrastructurev1alpha2.AWSControlPlaneList
+
+		err := h.k8sClient.CtrlClient().List(
+			ctx,
+			&list,
+			client.InNamespace(cr.GetNamespace()),
+			client.MatchingLabels{label.Cluster: key.ClusterID(cr)},
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(list.Items) == 0 {
+			return microerror.Mask(notFoundError)
+		}
+		if len(list.Items) > 1 {
+			return microerror.Mask(tooManyCRsError)
+		}
+
+		aws = list.Items[0]
+	}
+
+	// When the state machine is initialized we need to reset all internal
+	// information, including the pointer. This ensures we start fresh all over
+	// again.
+	{
+		h.ptr = 0
+	}
+
+	// We need a deterministic list of availability zones which we can cycle
+	// through for the required amount of masters. Eventually it happens that
+	// there is only 1 availability zone in a HA Masters setup. Therefore the
+	// internal state holds the given availability zones repeatedly 3 times so
+	// that it can always guarantee an availability zone for each master in any of
+	// the allowed permutations. Note that given 3 availability zones in a HA
+	// Masters setup will never make use of the repeated availability zones, which
+	// is just a side effect of the current implementation.
+	{
+		h.azs = []string{}
+
+		h.azs = append(h.azs, aws.Spec.AvailabilityZones...)
+		h.azs = append(h.azs, aws.Spec.AvailabilityZones...)
+		h.azs = append(h.azs, aws.Spec.AvailabilityZones...)
+	}
+
+	// The master IDs are only allowed to be either 0, 1, 2 or 3, so we hard code
+	// the list of master IDs depending on the given replicas, which must be
+	// either 1 or 3.
+	{
+		h.ids = []int{}
+
+		if key.G8sControlPlaneReplicas(g8s) == 1 {
+			h.ids = append(h.ids, 0)
+		}
+		if key.G8sControlPlaneReplicas(g8s) == 3 {
+			h.ids = append(h.ids, 1)
```

Comment:
this might cause an issue ai see in the templating 

when there is single master the resource for the first master is without number if there is  ha masters than the resource has an index 1, this can cause a potential issue during an upgrade from single to HA masters as it will cause deletion of the other resource as the name differs

that maybe fine for some resources for it does not work well with resource like etcd volume as that would  create a new empty volume

Reply:
In this case the problem is not with the state machine but the key functions which should then simply ignore the suffix, which most of them do already.

Reply:
We can do that as well, returning string without index on id 0 or id 1